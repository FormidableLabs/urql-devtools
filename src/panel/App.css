@import url("https://fonts.googleapis.com/css?family=Roboto:400,500,700&display=swap");
@import url("https://fonts.googleapis.com/css?family=Roboto+Mono:400,500,700&display=swap");

body {
  font-family: "Roboto", sans-serif;
  font-weight: 300;
}

code,
pre {
  font-family: "Roboto Mono", monospace;
}

::-webkit-scrollbar {
  display: none;
}
