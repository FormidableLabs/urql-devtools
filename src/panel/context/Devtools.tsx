import { DevtoolsExchangeOutgoingMessage } from "@urql/devtools";
import React, {
  createContext,
  useEffect,
  FC,
  useRef,
  useCallback,
  useState,
  useContext
} from "react";
import { DevtoolsPanelConnectionName, PanelOutgoingMessage } from "../../types";

export interface DevtoolsContextType {
  sendMessage: (message: PanelOutgoingMessage) => void;
  addMessageHandler: (
    cb: (message: DevtoolsExchangeOutgoingMessage) => void
  ) => () => void;
  clientConnected: boolean;
}

export const DevtoolsContext = createContext<DevtoolsContextType>(null as any);

export const useDevtoolsContext = () => useContext(DevtoolsContext);

export const DevtoolsProvider: FC = ({ children }) => {
  const [clientConnected, setClientConnected] = useState(false);
  const connection = useRef(
    chrome.runtime.connect({ name: DevtoolsPanelConnectionName })
  );

  /** Collection of operation events */
  const messageHandlers = useRef<
    Record<string, (msg: DevtoolsExchangeOutgoingMessage) => void>
  >({});

  const sendMessage = useCallback<DevtoolsContextType["sendMessage"]>(
    msg => connection.current.postMessage(msg),
    []
  );

  const addMessageHandler = useCallback<
    DevtoolsContextType["addMessageHandler"]
  >(callback => {
    const i = index++;
    messageHandlers.current[i] = callback;

    return () => {
      delete messageHandlers.current[i];
    };
  }, []);

  useEffect(() => {
    // Relay the tab ID to the background page
    connection.current.postMessage({
      type: "init",
      tabId: chrome.devtools.inspectedWindow.tabId
    });

    const handleMessage = (msg: DevtoolsExchangeOutgoingMessage) => {
      return Object.values(messageHandlers.current).forEach(h => h(msg));
    };

    connection.current.onMessage.addListener(handleMessage);
    return () => connection.current.onMessage.removeListener(handleMessage);
  }, []);

  // Listen for client init connection
  useEffect(() => {
    addMessageHandler(message => {
      if (message.type === "init") {
        setClientConnected(true);
      } else if (message.type === "disconnect") {
        setClientConnected(false);
      }
    });
  }, [addMessageHandler, setClientConnected]);

  return (
    <DevtoolsContext.Provider
      value={{ sendMessage, addMessageHandler, clientConnected }}
    >
      {children}
    </DevtoolsContext.Provider>
  );
};

let index = 0;
