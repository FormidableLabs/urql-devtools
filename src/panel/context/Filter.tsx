import React, { createContext, FC, useState } from "react";

import { UrqlEvent } from "../../types";

export enum FilterType {
  Name = "name",
  Key = "key",
  Info = "info"
}

interface Filter {
  value: string;
  propName: FilterType;
  propGetter: (e: UrqlEvent) => string | number;
}

interface FilterContextValue {
  filters: Filter[];
  addFilter: (filter: Filter) => void;
  removeFilter: (filter: Filter) => void;
}

export const FilterContext = createContext<FilterContextValue>(
  undefined as any
);

function filterFilters(filters: Filter[], incomingFilter: Filter) {
  return filters.filter(
    f =>
      f.value !== incomingFilter.value && f.propName !== incomingFilter.propName
  );
}

export const FilterProvider: FC = ({ children }) => {
  const [filters, setFilters] = useState<Filter[]>([]);

  const addFilter = (filter: Filter) => {
    return setFilters([filter, ...filterFilters(filters, filter)]);
  };

  const removeFilter = (filter: Filter) => {
    return setFilters(filterFilters(filters, filter));
  };

  const value = {
    filters,
    addFilter,
    removeFilter
  };

  return (
    <FilterContext.Provider value={value}>{children}</FilterContext.Provider>
  );
};
