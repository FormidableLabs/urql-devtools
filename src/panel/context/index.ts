export * from "./Devtools";
export * from "./Events";
export * from "./Request";
export { ExplorerContext, ExplorerContextProvider } from "./Explorer";
