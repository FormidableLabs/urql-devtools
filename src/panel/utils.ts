import { DevtoolsExchangeOutgoingMessage } from "@urql/devtools";

const filterEventTypes = <T extends DevtoolsExchangeOutgoingMessage["type"]>(
  types: T[]
) => {};

type FilterType<T, A> = A extends { type: T } ? true : false;

const excludeEventTypes = <T extends DevtoolsExchangeOutgoingMessage["type"]>(
  types: T[]
) => <E extends DevtoolsExchangeOutgoingMessage>(
  event: E
): FilterType<T, E> => {
  if (types.indexOf(event.type as any)) {
    return true;
  }

  return false;
};
