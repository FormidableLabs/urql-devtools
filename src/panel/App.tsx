import "./App.css";
import React from "react";
import { HashRouter, Route, Redirect } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { Events } from "./events";
import { Explorer } from "./explorer";
import { Navigation } from "./Navigation";
import { Request } from "./request/Request";
import { theme } from "./theme";

import {
  DevtoolsProvider,
  EventsProvider,
  RequestProvider,
  ExplorerContextProvider
} from "./context";

export const App = () => {
  return (
    <DevtoolsProvider>
      <ThemeProvider theme={theme}>
        <HashRouter>
          <EventsProvider>
            <Route path="/events" component={Events} />
          </EventsProvider>
          <RequestProvider>
            <Route path="/request" component={Request} />
          </RequestProvider>
          <ExplorerContextProvider>
            <Route path="/explorer" exact component={Explorer} />
          </ExplorerContextProvider>
          <Route path="/" exact component={() => <Redirect to="/explorer" />} />
          <Navigation />
        </HashRouter>
      </ThemeProvider>
    </DevtoolsProvider>
  );
};
