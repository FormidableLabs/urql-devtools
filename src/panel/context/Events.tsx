import { print } from "graphql";
import React, {
  createContext,
  FC,
  useState,
  useEffect,
  useContext,
  useMemo,
  useCallback
} from "react";
import {
  ParsedEvent,
  ParsedMutationEvent,
  ParsedQueryEvent,
  ParsedResponseEvent,
  ParsedErrorEvent,
  EventType,
  ParsedSubscriptionEvent,
  ParsedTeardownEvent
} from "../types";
import {
  UrqlEvent,
  OutgoingOperation,
  IncomingResponse,
  IncomingError
} from "../../types-old";
import { DevtoolsContext } from "./Devtools";
import {
  DevtoolsExchangeOutgoingMessage,
  OperationMessage,
  InitMessage
} from "@urql/devtools";

export interface EventsContextValue {
  events: ParsedEvent[];
  selectedEvent?: ParsedEvent;
  selectEvent: (event: ParsedEvent) => void;
  clearSelectedEvent: () => void;
  activeFilters: FilterState;
  addFilter: <T extends keyof FilterState, V extends FilterState[T][number]>(
    property: T,
    value: V
  ) => void;
  removeFilter: <T extends keyof FilterState, V extends FilterState[T][number]>(
    property: T,
    value: V
  ) => void;
}

interface FilterState {
  type: EventType[];
  source: string[];
  key: number[];
}

type PresentedEvent = Exclude<DevtoolsExchangeOutgoingMessage, InitMessage>;

export const EventsContext = createContext<EventsContextValue>(null as any);

export const EventsProvider: FC = ({ children }) => {
  const { addMessageHandler } = useContext(DevtoolsContext);
  const [rawEvents, setRawEvents] = useState<PresentedEvent[]>([]);
  const [selectedEvent, setSelectedEvent] = useState<ParsedEvent | undefined>(
    undefined
  );
  const [filters, setFilters] = useState<FilterState>({
    type: [],
    source: [],
    key: []
  });

  // /** Set initial state from cache */
  // useEffect(() => {
  //   window.chrome.devtools.inspectedWindow.eval(
  //     `window.__urql__.events`,
  //     (ops: UrqlEvent[]) => {
  //       console.log(ops);
  //       setRawEvents(ops);
  //     }
  //   );
  // }, []);

  /** Handle incoming events */
  useEffect(() => {
    return addMessageHandler(msg => {
      if (!["operation", "response", "error"].includes(msg.type)) {
        return;
      }

      setRawEvents(o => [msg as PresentedEvent, ...o]);
    });
  }, []);

  const addFilter = useCallback<EventsContextValue["addFilter"]>(
    (property, value) =>
      setFilters(f => ({
        ...f,
        [property]: [...new Set([...f[property], value])]
      })),
    []
  );

  const removeFilter = useCallback<EventsContextValue["removeFilter"]>(
    (property, value) =>
      setFilters(f => ({
        ...f,
        [property]: (f[property] as typeof value[]).filter(val => val !== value)
      })),
    []
  );

  const selectEvent = useCallback(
    (op: ParsedEvent) => setSelectedEvent(op),
    []
  );

  const clearSelectedEvent = useCallback(() => setSelectedEvent(undefined), []);

  const applyFilter = useCallback(
    (e: ParsedEvent) => {
      const typeFilter =
        filters.type.length === 0 ? true : filters.type.includes(e.type);
      const sourceFilter =
        filters.source.length === 0 ? true : filters.source.includes(e.source);
      const keyFilter =
        filters.key.length === 0 ? true : filters.key.includes(e.key);

      return typeFilter && sourceFilter && keyFilter;
    },
    [filters]
  );

  /** Events which are cleaned for rendering */
  const events = useMemo(
    () =>
      rawEvents
        .map(e =>
          e.type === "operation"
            ? parseOperation(rawEvents, e)
            : parseResponse(e)
        )
        .filter(applyFilter),
    [applyFilter, rawEvents]
  );

  const value = {
    events,
    selectedEvent,
    selectEvent,
    clearSelectedEvent,
    activeFilters: filters,
    addFilter,
    removeFilter
  };

  return <EventsContext.Provider value={value} children={children} />;
};

const parseOperation = (
  allEvents: PresentedEvent[],
  event: OutgoingOperation
):
  | ParsedQueryEvent
  | ParsedMutationEvent
  | ParsedSubscriptionEvent
  | ParsedTeardownEvent => {
  const shared = {
    key: event.data.key,
    source: (event.data.context.meta as any).source,
    timestamp: event.timestamp
  } as const;
  const queryPanel = { name: "query", data: print(event.data.query) } as const;
  const varsPanel = { name: "variables", data: event.data.variables } as const;
  const metaPanel = {
    name: "meta",
    data: {
      key: event.data.key,
      operationName: event.data.operationName,
      context: event.data.context
    }
  } as const;

  const type = event.data.operationName;

  const responseEvent = allEvents
    // Only events after the request
    .filter(e => e.type !== "operation" && e.timestamp > event.timestamp)
    // First response for mutation, latest response for query
    .sort((a, b) =>
      type === "mutation"
        ? a.timestamp - b.timestamp
        : b.timestamp - a.timestamp
    )
    .find(
      e =>
        (e as IncomingError | IncomingResponse).data.operation.key ===
        event.data.key
    ) as IncomingError | IncomingResponse | undefined;
  const responseData =
    responseEvent === undefined
      ? {}
      : { data: responseEvent.data.data, error: responseEvent.data.error };

  /**
   * Conditionals can't be called during object construction
   * if we want to keep implicit type safety :(
   */
  if (type === "mutation") {
    return {
      type,
      ...shared,
      panels: [
        queryPanel,
        varsPanel,
        { name: "response", data: responseData },
        metaPanel
      ]
    };
  }

  if (type === "query") {
    return {
      type,
      ...shared,
      panels: [
        queryPanel,
        varsPanel,
        { name: "state", data: responseData },
        metaPanel
      ]
    };
  }

  return {
    type,
    ...shared,
    panels: [metaPanel]
  };
};

const parseResponse = (
  event: IncomingResponse | IncomingError
): ParsedResponseEvent | ParsedErrorEvent => {
  const shared = {
    key: event.data.operation.key,
    source: (event.data.operation.context.meta as any).source,
    timestamp: event.timestamp
  };
  const metaPanel = {
    name: "meta",
    data: {
      key: event.data.operation.key,
      operationName: event.data.operation.operationName,
      context: event.data.operation.context
    }
  } as const;

  const type = event.data.error !== undefined ? "error" : ("response" as const);

  if (type === "error") {
    return {
      type,
      ...shared,
      panels: [{ name: "error", data: event.data.error }, metaPanel]
    };
  }

  // TODO: Fix
  if (type === "response" || true) {
    return {
      type,
      ...shared,
      panels: [{ name: "response", data: event.data.data }, metaPanel]
    };
  }
};
