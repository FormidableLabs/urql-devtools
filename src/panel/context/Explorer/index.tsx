import React, {
  createContext,
  useState,
  useEffect,
  FC,
  useMemo,
  SetStateAction,
  Dispatch,
} from "react";
import { useDevtoolsContext } from "../Devtools";
import { handleResponse, ParsedNodeMap, ParsedFieldNode } from "./ast";

export interface ExplorerContextValue {
  operations: ParsedNodeMap;
  expandedNodes: ParsedFieldNode[];
  setExpandedNodes: Dispatch<SetStateAction<ParsedFieldNode[]>>;
  focusedNode?: ParsedFieldNode;
  setFocusedNode: Dispatch<SetStateAction<ParsedFieldNode | undefined>>;
}

export const ExplorerContext = createContext<ExplorerContextValue>(null as any);

export const ExplorerProvider: FC = ({ children }) => {
  const { addMessageHandler } = useDevtoolsContext();
  const [operations, setOperations] = useState<
    ExplorerContextValue["operations"]
  >({});
  const [expandedNodes, setExpandedNodes] = useState<
    ExplorerContextValue["expandedNodes"]
  >([]);
  const [focusedNode, setFocusedNode] = useState<
    ExplorerContextValue["focusedNode"]
  >(undefined);

  useEffect(() => {
    return addMessageHandler((message) => {
      if (message.type === "connection-disconnect") {
        setOperations({});
        return;
      }

      if (message.type !== "debug-event" || message.data.type !== "update") {
        return;
      }

      const debugEvent = message.data;

      if (debugEvent.data) {
        setOperations((operations) =>
          handleResponse({
            operation: debugEvent.operation,
            data: debugEvent.data.value,
            parsedNodes: operations,
          })
        );
        return;
      }
    });
  }, [addMessageHandler]);

  const value = useMemo(
    () => ({
      expandedNodes,
      setExpandedNodes,
      focusedNode,
      setFocusedNode,
      operations,
    }),
    [operations, focusedNode, setFocusedNode, expandedNodes, setExpandedNodes]
  );

  return (
    <ExplorerContext.Provider value={value}>
      {children}
    </ExplorerContext.Provider>
  );
};
