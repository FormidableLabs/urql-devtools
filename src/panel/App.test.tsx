jest.mock("./context/Devtools.tsx", () => {
  return {
    ...(jest.requireActual("./context/Devtools.tsx") as Record<
      string,
      unknown
    >),
    useDevtoolsContext: jest.fn(),
  };
});
import React from "react";
import { shallow } from "enzyme";
import { mocked } from "ts-jest/utils";
import { App, AppRoutes } from "./App";
import { useDevtoolsContext } from "./context";

describe("App", () => {
  describe("on mount", () => {
    it("matches snapshot", () => {
      expect(shallow(<App />)).toMatchSnapshot();
    });
  });
});

describe("App routes", () => {
  describe("on mount", () => {
    describe("on connected", () => {
      beforeEach(() => {
        mocked(useDevtoolsContext).mockReturnValue({
          client: {
            connected: true,
            version: {
              mismatch: false,
            },
          },
        } as any);
      });

      it("matches snapshot", () => {
        expect(shallow(<AppRoutes />)).toMatchSnapshot();
      });
    });

    describe("on version mismatch", () => {
      beforeEach(() => {
        mocked(useDevtoolsContext).mockReturnValue({
          client: {
            connected: true,
            version: {
              required: "9.9.9",
              actual: "0.0.1",
              mismatch: true,
            },
          },
        } as any);
      });

      it("matches snapshot", () => {
        expect(shallow(<AppRoutes />)).toMatchSnapshot();
      });
    });

    describe("on disconnected", () => {
      beforeEach(() => {
        mocked(useDevtoolsContext).mockReturnValue({
          client: {
            connected: false,
          },
        } as any);
      });

      it("matches snapshot", () => {
        expect(shallow(<AppRoutes />)).toMatchSnapshot();
      });
    });
  });
});
