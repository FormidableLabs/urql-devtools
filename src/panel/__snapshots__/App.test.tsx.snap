// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App on mount matches snapshot 1`] = `
<ThemeProvider
  theme={
    Object {
      "blue": Object {
        "+1": "#1a90ff",
        "+2": "#339dff",
        "+3": "#4da9ff",
        "+4": "#66b5ff",
        "+5": "#80c1ff",
        "+6": "#99ceff",
        "+7": "#b3daff",
        "+8": "#cce6ff",
        "+9": "#e5f3ff",
        "-1": "#0077e6",
        "-2": "#006acc",
        "-3": "#005cb3",
        "-4": "#004f99",
        "-5": "#004280",
        "-6": "#003566",
        "-7": "#00284d",
        "-8": "#001a33",
        "-9": "#000d19",
        "0": "#0084FF",
      },
      "dark": Object {
        "+1": "#050505",
        "+2": "#0a0a0a",
        "+3": "#0f0f0f",
        "+4": "#141414",
        "+5": "#1a1a1a",
        "+6": "#1f1f1f",
        "+7": "#242424",
        "+8": "#292929",
        "+9": "#2e2e2e",
        "0": "#000",
      },
      "green": Object {
        "+1": "#92e37d",
        "+2": "#a3e892",
        "+3": "#b5eca7",
        "+4": "#c7f1bc",
        "+5": "#d8f5d1",
        "+6": "#eafae6",
        "+7": "#fcfefb",
        "+8": "#fff",
        "+9": "#fff",
        "-1": "#6edb53",
        "-2": "#5dd63e",
        "-3": "#4dcf2c",
        "-4": "#45ba27",
        "-5": "#3da523",
        "-6": "#35901e",
        "-7": "#2d7b1a",
        "-8": "#266516",
        "-9": "#1e5011",
        "0": "#80DF68",
      },
      "grey": Object {
        "+1": "#858585",
        "+2": "#8a8a8a",
        "+3": "#8f8f8f",
        "+4": "#949494",
        "+5": "#9a9a9a",
        "+6": "#9f9f9f",
        "+7": "#a4a4a4",
        "+8": "#a9a9a9",
        "+9": "#aeaeae",
        "-1": "#7b7b7b",
        "-2": "#767676",
        "-3": "#717171",
        "-4": "#6c6c6c",
        "-5": "#676767",
        "-6": "#616161",
        "-7": "#5c5c5c",
        "-8": "#575757",
        "-9": "#525252",
        "0": "#808080",
      },
      "light": Object {
        "-1": "#fafafa",
        "-2": "#f5f5f5",
        "-3": "#f0f0f0",
        "-4": "#ebebeb",
        "-5": "#e6e6e6",
        "-6": "#e0e0e0",
        "-7": "#dbdbdb",
        "-8": "#d6d6d6",
        "-9": "#d1d1d1",
        "0": "#fff",
      },
      "orange": Object {
        "+1": "#ff8b5a",
        "+2": "#ff9d73",
        "+3": "#ffaf8d",
        "+4": "#ffc1a6",
        "+5": "#ffd2c0",
        "+6": "#ffe4d9",
        "+7": "#fff6f3",
        "+8": "#fff",
        "+9": "#fff",
        "-1": "#ff6726",
        "-2": "#ff550d",
        "-3": "#f34800",
        "-4": "#d94100",
        "-5": "#c03900",
        "-6": "#a63200",
        "-7": "#8d2a00",
        "-8": "#732200",
        "-9": "#5a1b00",
        "0": "#FF7940",
      },
      "pink": Object {
        "+1": "#fb7ec8",
        "+2": "#fc96d3",
        "+3": "#fcafdd",
        "+4": "#fdc8e8",
        "+5": "#fee0f2",
        "+6": "#fff9fd",
        "+7": "#fff",
        "+8": "#fff",
        "+9": "#fff",
        "-1": "#f94cb4",
        "-2": "#f834a9",
        "-3": "#f81b9f",
        "-4": "#f10893",
        "-5": "#d80784",
        "-6": "#c00675",
        "-7": "#a70566",
        "-8": "#8e0557",
        "-9": "#760448",
        "0": "#FA65BE",
      },
      "purple": Object {
        "+1": "#9053f4",
        "+2": "#a06bf5",
        "+3": "#af83f7",
        "+4": "#be9bf8",
        "+5": "#ceb3fa",
        "+6": "#ddcafc",
        "+7": "#ede2fd",
        "+8": "#fcfaff",
        "+9": "#fff",
        "-1": "#7223f0",
        "-2": "#6310ea",
        "-3": "#590ed3",
        "-4": "#4f0cbb",
        "-5": "#450ba3",
        "-6": "#3b098b",
        "-7": "#310873",
        "-8": "#27065b",
        "-9": "#1c0443",
        "0": "#813BF2",
      },
      "red": Object {
        "+1": "#ff6770",
        "+2": "#ff8088",
        "+3": "#ff9aa0",
        "+4": "#ffb3b8",
        "+5": "#ffcdd0",
        "+6": "#ffe6e8",
        "+7": "#fff",
        "+8": "#fff",
        "+9": "#fff",
        "-1": "#ff3340",
        "-2": "#ff1a28",
        "-3": "#ff0010",
        "-4": "#e6000e",
        "-5": "#cd000d",
        "-6": "#b3000b",
        "-7": "#9a0009",
        "-8": "#800008",
        "-9": "#670006",
        "0": "#FF4D58",
      },
    }
  }
>
  <DevtoolsProvider>
    <AppRoutes />
  </DevtoolsProvider>
</ThemeProvider>
`;

exports[`App routes on mount on connected matches snapshot 1`] = `
<HashRouter>
  <TimelineProvider>
    <Route
      component={[Function]}
      path="/events"
    />
  </TimelineProvider>
  <RequestProvider>
    <Route
      component={[Function]}
      path="/request"
    />
  </RequestProvider>
  <ExplorerProvider>
    <Route
      component={[Function]}
      exact={true}
      path="/explorer"
    />
  </ExplorerProvider>
  <Route
    component={[Function]}
    exact={true}
    path="/"
  />
  <Navigation />
</HashRouter>
`;

exports[`App routes on mount on disconnected matches snapshot 1`] = `<Disconnected />`;
